### PR TITLE
Update range for remapped sparse textures instead of recreating them

### PR DESCRIPTION
**Pull request by riperiperi**
Sparsely mapped textures are useful when a game wants texture data to cover a large area (say, an entire level), but loading the whole thing would take up too much memory. You can load and unload parts of the texture by mapping physical pages into the texture's virtual memory allocation in GPU address space, and unmapping them when you're finished. This way, you only pay for the pages you use.

Ryujinx's way of handling sparsely mapped textures was to delete their pool references entirely when _any_ mapping change happened, as the texture cache would simply recreate it the next time it was used. This works, but is slow, especially if your game has mapped several 48MB textures in this way, and changes mappings mid-game.

This PR adds the ability to migrate a texture to a new set of physical ranges. The existing texture data will remain intact, and any regions where the mapping has changed will be marked as dirty and synchronized on use. Note that it will only happen for the earliest GPU VA pool reference on the texture, any other pool references will be deleted as before.

Greatly improves stuttering going through doors on Metroid Prime: Remastered. Does not eliminate it entirely, because the game uses ASTC textures, which must be decoded before uploading on desktop, and are loaded around the same time a door is opened.

## Other fixes

This PR also fixes a few issues with partial updates and subresource tracking:
- Fixed an issue where handles for 3d texture slices would have a very small size, as it would only count to the beginning of the next slice (which may have been interleaved with this one)
- Partial updates are now enabled when using a texture and _any_ subresource is not modified, rather than only when a handle is unmapped. May improve performance in games with texture streaming.
- Subresource tracking (layers/levels) are now enabled on sufficiently large 3D and array textures to avoid impacts from synchronizing the whole thing. They are also enabled if the texture is partially mapped. This reduces stuttering from texture updates.
  - Greatly reduces FIFO% in NieR: Automata when travelling between areas. Framerate still dips due to Vulkan buffer memory stuff and cb0 accesses that can't be eliminated.
  - May affect texture streaming performance in Xenoblade. Who knows in what way.
- Gets blocks of slices from the source texture rather than one slice at a time, so that 3D textures don't copy the same interleaved data range each slice (potentially up to 16x the data)
- Partial uploads of 3D textures no longer allocate arrays for the full depth of the texture instead of one slice.

## Areas for improvement

A few things could still be improved:
- Handles are fully recreated every time a sparse mapping is changed, not just the handles that overlapped the ranges that changed mapping. Improving this could be very difficult.
- InheritHandles and ReplaceHandles are `O(n*m)` _best case_. This was the case for InheritHandles before, but it is more of a problem with sparse textures as they could be remapped whenever and the game assumes it's free.
- Texture ranges can't be updated if they have any views, are a view themselves, or have any form of copy dependency. This is because the texture on the other side could have a pool reference with a different GPU VA. I'm not sure if there's a way to handle this, or if there's any need to. It's very unlikely that a sparsely mapped texture will be rendered to, and that's typically what creates views. There's always swizzle and format alias, though...
- The backend could be made aware of layers/levels that are fully unmapped and do sparse mapping for real, though this could get quite complicated.

NOTE: draft for some small self-review cleanup and hopefully game testing.

This changes quite a lot regarding unmapped textures and data uploads, so it would be a good idea to test this in a variety of games.

### CODE REVIEW

**Comment by Bjorn29512** [comment]
Tested 20+ titles and invested a couple minutes of playtime in each, noticed no performance or any visual difference compared to master, cant speak for specific scenarios that would break any visuals or any crashes introduced since that is outside the scope of a limited testing like this. Metroid prime did get a significant improvement when it comes stuttering as per the objective of the PR.

Attachments showing the titles tested : 
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image
image

**Comment by Squall-Leonhart** [comment]
This significantly improves the issue reported in #4420

**Comment by Bjorn29512** [comment]
> This improves the issue reported in #4420 but Fifo still eventually hits 99%

No its still the same, no such effect you are talking about.

image

File: Ryujinx.Graphics.Gpu/Image/TextureCache.cs [hunk 1]
```diff
@@ -194,6 +195,39 @@ public void Lift(Texture texture)
...
+        {
+            // There cannot be an existing texture compatible with this mapping in the texture cache already.
+            int overlapCount = _textures.FindOverlaps(range, ref _textureOverlaps);
+
+            for (int i = 0; i < overlapCount; i++)
+            {
+                var other = _textureOverlaps[i];
+                
+                if (texture != other &&
+                    (texture.IsViewCompatible(other.Info, other.Range, true, other.LayerSize, _context.Capabilities, out int _, out int _) != TextureViewCompatibility.Incompatible ||
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureCache.cs:215]:
```suggestion
                    (texture.IsViewCompatible(other.Info, other.Range, true, other.LayerSize, _context.Capabilities, out _, out _) != TextureViewCompatibility.Incompatible ||
```

File: Ryujinx.Graphics.Gpu/Image/TextureCache.cs [hunk 2]
```diff
@@ -194,6 +195,39 @@ public void Lift(Texture texture)
...
+                    other.IsViewCompatible(texture.Info, texture.Range, true, texture.LayerSize, _context.Capabilities, out int _, out int _) != TextureViewCompatibility.Incompatible))
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureCache.cs:216]:
```suggestion
                    other.IsViewCompatible(texture.Info, texture.Range, true, texture.LayerSize, _context.Capabilities, out _, out _) != TextureViewCompatibility.Incompatible))
```

File: Ryujinx.Graphics.Gpu/Image/TextureGroup.cs [hunk 3]
```diff
@@ -39,6 +39,11 @@ public TextureIncompatibleOverlap(TextureGroup group, TextureViewCompatibility c
     /// </summary>
     class TextureGroup : IDisposable
     {
+        /// <summary>
+        /// Threshold of layers to force granular handles (and thus partial loading) on array/3d textures.
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:43]:
```suggestion
        /// Threshold of layers to force granular handles (and thus partial loading) on array/3D textures.
```

File: Ryujinx.Graphics.Gpu/Image/TextureGroup.cs [hunk 4]
```diff
@@ -331,24 +359,55 @@ public void SynchronizeMemory(Texture texture)
...
                 {
                     var info = GetHandleInformation(baseHandle + i);
 
+                    if (spanEndIndex <= i - 1)
+                    {
+                        spanEndIndex = i;
+
+                        if (_is3D)
+                        {
+                            // Look ahead to see how many handles need loaded.
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:378]:
```suggestion
                            // Look ahead to see how many handles need loaded.
```
"need loaded" sounds bit weird. Maybe "needs to be loaded"?

File: Ryujinx.Graphics.Gpu/Image/TexturePool.cs [hunk 5]
```diff
@@ -1,19 +1,45 @@
...
+    /// <summary>
+    /// A request to dereference a texture from a pool.
+    /// </summary>
+    struct DereferenceRequest
+    {
+        public readonly bool IsUnmapped;
+        public readonly Texture Texture;
+        public readonly int ID;
+
+        public DereferenceRequest(Texture texture)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TexturePool.cs:21]: Missing XML docs for the constructors and fields.

File: Ryujinx.Graphics.Gpu/Image/TexturePool.cs [hunk 6]
```diff
@@ -159,24 +189,95 @@ public void ForceRemove(Texture texture, int id, bool deferred)
...
+
+                    Items[request.ID] = texture;
+
+                    // Create a new pool reference, as the last one was removed on unmap.
+
+                    texture.IncrementReferenceCount(this, request.ID, address);
+                    texture.DecrementReferenceCount();
+
+                    // Refetch the range. Changes since the last check could have been lost
+                    // as the cache entry was not restored (required to queue mapping change)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TexturePool.cs:253]:
```suggestion
                    // as the cache entry was not restored (required to queue mapping change).
```

**Comment by ActualMandM** [comment]
No noticeable difference with Groove Coaster and Project DIVA between master and this PR.

**Comment by riperiperi** [comment]
I think I'm done tweaking this, should probably retest to make sure everything's still working.

File: Ryujinx.Graphics.Gpu/Image/TextureGroup.cs [hunk 7]
```diff
@@ -7,6 +7,7 @@
 using Ryujinx.Memory.Range;
 using System;
 using System.Collections.Generic;
+using System.Reflection.Metadata;
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:10]: Why is that needed?

File: Ryujinx.Graphics.Gpu/Image/TextureGroup.cs [hunk 8]
```diff
@@ -1057,17 +1121,55 @@ public void Inherit(TextureGroup other)
...
                 {
-                    foreach (CpuRegionHandle handle in groupHandle.Handles)
+                    if (rangeChanged)
                     {
-                        handle.Reprotect();
+                        // When the storage range changes, this becomes a little different.
+                        // If a range does not match one in the original, treat it as modified.
+                        // It has been newly mapped and its data must be synchronized.
+
+                        if (groupHandle.Handles.Length != 0)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:1139]: Would be nice to change it to something like:
```cs
if (groupHandle.Handles.Length == 0)
{
    continue;
}
```
To avoid all this nesting.

File: Ryujinx.Graphics.Gpu/Image/TextureGroup.cs [hunk 9]
```diff
@@ -1057,17 +1121,55 @@ public void Inherit(TextureGroup other)
...
+                        {
+                            foreach (var oldGroup in _handles)
+                            {
+                                if (groupHandle.OverlapsWith(oldGroup.Offset, oldGroup.Size))
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TextureGroup.cs:1143]: Similar suggestion to the above to avoid nesting.

File: Ryujinx.Graphics.Gpu/Image/TexturePool.cs [hunk 10]
```diff
@@ -12,8 +15,61 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            {
+                IsRemapped = isRemapped;
+                Texture = texture;
+                ID = id;
+            }
+
+            /// <summary>
+            /// Create a dereference request for a texture removal.
+            /// </summary>
+            /// <param name="texture">The texture being removed</param>
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TexturePool.cs:54]: Missing `<returns>`.

File: Ryujinx.Graphics.Gpu/Image/TexturePool.cs [hunk 11]
```diff
@@ -12,8 +15,61 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            public static DereferenceRequest Remove(Texture texture)
+            {
+                return new DereferenceRequest(false, texture, 0);
+            }
+
+            /// <summary>
+            /// Create a dereference request for a texture remapping with a specific pool ID.
+            /// </summary>
+            /// <param name="texture">The texture being remapped</param>
+            /// <param name="id">The ID of the pool entry, used to restore remapped textures</param>
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TexturePool.cs:64]: Missing `<returns>`.

File: Ryujinx.Graphics.Gpu/Image/TexturePool.cs [hunk 12]
```diff
@@ -155,28 +215,106 @@ public int CheckModified()
...
 
+        /// <summary>
+        /// Queues a request to update a texture's mapping. 
+        /// Mapping is updated later to avoid deleting the texture if it is still sparsely mapped.
+        /// </summary>
+        /// <param name="texture">Texture with potential mapping change</param>
+        /// <param name="id">ID in cache of texture with potential mapping change</param>
+        public void QueueUpdateMapping(Texture texture, int id)
+        {
+            if (Interlocked.Exchange(ref Items[id], null) != null)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/TexturePool.cs:241]: Am I correct assuming that `Items[id]` is expected to be the same as `texture` if not null? If so, I think it's possible to remove the texture parameter and use the previous value on `Items[id]`.

File: Ryujinx.Graphics.Gpu/Image/Texture.cs [hunk 13]
```diff
@@ -1585,6 +1605,36 @@ public void RemoveFromPools(bool deferred)
...
+                {
+                    if (address == 0 || address == owner.GpuAddress)
+                    {
+                        address = owner.GpuAddress;
+
+                        owner.Pool.QueueUpdateMapping(this, owner.ID);
+                    }
+                    else
+                    {
+                        // If there is a different GPU va mapping, prefer the first and delete the others.
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/Texture.cs:1627]:
```suggestion
                        // If there is a different GPU VA mapping, prefer the first and delete the others.
```

**Review by gdkchan — APPROVED** [review]
lgtm, thanks. Would be nice to address the remaining comment but it's not a big deal.

File: Ryujinx.Graphics.Gpu/Image/TexturePool.cs (hunk 12 above)
Comment (riperiperi) [reply Ryujinx.Graphics.Gpu/Image/TexturePool.cs:243]: My initial worry is that `Items[id]` could be swapped for another non-null texture, but I've spent a long time figuring out a case where that might happen, and cannot find one (due to the combination of `lock` and interlocked exchange). This method is always called within `_poolOwners` lock, which greatly limits how much can change. 

I can change it, though maybe to be extra cautious, it should instead be an exchange check to see if the old value is equal to `texture`.

**Comment by Khobztajine** [comment]
Hi am new on github and I want to thank you for this build, On xenoblade chronicle definitive edition, I had 35/55 fps with a lot of stutter with the main version of ryujinx but with this build my game run at 60 fps with no drop and no stutter, I dont know what magic it is but its work amazingly.

**Comment by Disorientator** [comment]
Sorry in advance if this is a stupid question, but does the ASTC texture decoding take place on the CPU or GPU?

**Comment by GamerzHell9137** [comment]
> Sorry in advance if this is a stupid question, but does the ASTC texture decoding take place on the CPU or GPU?

CPU, and no, changing to GPU wouldn't help much.

**Comment by Disorientator** [comment]
Yeap, that was my next question. Thanks for clarifying.
